iaso/permissions.py

```diff
@@ -1,8 +1,30 @@
 from rest_framework import permissions
+from rest_framework.exceptions import NotAuthenticated
+
+from iaso.api.serializers import AppIdSerializer
+from iaso.models import Project
 
 
 class ReadOnly(permissions.BasePermission):
     def has_permission(self, request, view):
         if request.method in permissions.SAFE_METHODS:
             return True
         return False
+
+
+class IsAuthenticatedOrReadOnlyWhenNoAuthenticationRequired(permissions.IsAuthenticatedOrReadOnly):
+    def has_permission(self, request, view):
+        app_id = AppIdSerializer(data=request.query_params).get_app_id(raise_exception=False)
+        if app_id is not None:
+            try:
+                project = Project.objects.get(app_id=app_id)
+                if not bool(request.user and request.user.is_authenticated):
+                    if project.needs_authentication:
+                        raise NotAuthenticated()
+                elif request.user.iaso_profile.account.id != project.account.id:
+                    raise NotAuthenticated()
+
+            except Project.DoesNotExist:
+                return super().has_permission(request, view)
+
+        return super().has_permission(request, view)
```

Comment:
This basically does exactly the same thing with slightly less code.

This is just a suggestion, there is absolutely no obligation to merge it.

```suggestion
class IsAuthenticatedOrReadOnlyWhenNoAuthenticationRequired(permissions.IsAuthenticatedOrReadOnly):
    def has_permission(self, request, view):
        app_id = AppIdSerializer(data=request.query_params).get_app_id(raise_exception=False)

        if app_id:
            try:
                project = Project.objects.get(app_id=app_id)
            except Project.DoesNotExist:
                return super().has_permission(request, view)

            if not request.user.is_authenticated and project.needs_authentication:
                raise NotAuthenticated()

            if request.user.iaso_profile.account.id != project.account.id:
                raise NotAuthenticated()

        return super().has_permission(request, view)
```

Reply:
I agree

Reply:
@kemar I believe I wrote it the way I did because, unless I'm mistaken, you will have a crash if the project doesn't require authentication, but the user is anonymous.
Indeed, in case the user is anonymous, and the project doesn't require authentication, the line:
```
if request.user.iaso_profile.account.id != project.account.id:
```
will crash with `user doesn't have an attribute 'iaso_profile'`, or am I missing something?

Reply:
@bmonjoie yup you're right, I went too fast but the idea was to remove a level of nesting to make it easier to read.

Reply:
I have pushed a rewrite of this class in an attempt to make it more readable.

Reply:
Thanx @bmonjoie indeed it's more readable! 🙇